Extensibility section proposal

sgillies opened the pull request:
On the GeoJSON email list, I introduced the idea of a 5-point extensibility section and here is the start of it. I'm going to add inline comments at a few sections, some mine, some from the list.

sgillies commented on considerations.mkd:
```diff
@@ -58,3 +58,38 @@ precision than necessary.
+In GeoJSON, an "extension" is any property not defined in section 2.
+Consuming implementations that encounter unfamiliar extensions MUST NOT
+stop processing or signal an error and MUST continue processing the
+items as if those properties were not present. Note that support for
```
This is precisely the same as in https://tools.ietf.org/html/rfc7493#section-4.2. Unrecognized extensions must be ignored, they must not cause the rest of the GeoJSON document to be rejected.

mpdaly replied:
That reference is definitely *not* about extensions: it is about future core protocol elements. There are no extensions in I-JSON.

sgillies replied:
@mpdaly I disagree. The first paragraph in RFC 7493 4.2 describes extensions. It definitely makes no "core" vs "non-core" distinction.

mpdaly replied:
> It is frequently the case that changes to protocols are required
> after they have been put in production.  Protocols that allow the
> introduction of new protocol elements in a way that does not disrupt
> the operation of existing software have proven advantageous in
> practice.

That paragraph that you reference quite specifically mentions "changes to [this] protocol", ie version 2, and quite specifically **does not** mention extensions. In fact, the word extension **does not appear** anywhere in the document.

sgillies commented on considerations.mkd:
```diff
@@ -58,3 +58,38 @@ precision than necessary.
+
+Implementations MUST NOT extend the fixed set of GeoJSON types:
+FeatureCollection, Feature, Point, LineString, MultiPoint, Polygon,
+MultiLineString, MultiPolygon, and GeometryCollection.
```
Examples:

- Implementations cannot add circles or curves. Only core geometry types are allowed.
- Implementations cannot add "Instant" or "Interval" types. Note that this kills our Time extension (elsewhere) as it is currently written.

sgillies commented on considerations.mkd:
```diff
@@ -58,3 +58,38 @@ precision than necessary.
+## Semantics of GeoJSON properties are fixed
+
+Implementations MUST NOT change the the semantics of GeoJSON properties.
+
```
I'm not sure about the right examples to add here, if any. As it's written, GeoJSON has the least semantics possible.

An extension like the following is one of the things we disallow:

```
{ "type": "MultiPoint", "coordinates": [[0, 0], [1, 1]], "is_a_circle": True }
```

dret replied:
On 2015-11-25 20:00, Sean Gillies wrote:
> In considerations.mkd
>> +## Semantics of GeoJSON properties are fixed
>> +Implementations MUST NOT change the the semantics of GeoJSON properties.
> I'm not sure about the right examples to add here, if any.

maybe something that would change the coordinate semantics? i think your 
example of introducing additional geometries is a good one.

JerrySievert replied:
if we are not going to allow extensions such as `"is_a_circle_not_a_multipoint"`, would there be consideration for adding `properties` or another private namespace inside of a GeoJSON types?

wrapping something in a Feature in order to add a single element can be very verbose.

the (contrived) case I am thinking of is:

```
{
  "type": "Point",
  "coordinates": [ 0, 0, 100 ],
  "moment": true
}
```

in this case, it would be encapsulating of WKT `POINT M (1 1 100)`.

sgillies replied:
@JerrySievert GeoJSON has no `Point M` type and so the above looks a lot to me like changing the semantics of `Point`. Something for the list to discuss, for sure.

JerrySievert replied:
i know that we discussed this on the geojson list a while back (both in the context of "time" in may 2013, and in the context of GPX).

in the case of `Point M` simply defines that the third coordinate is "moment" - i don't see that as changing the semantics of `Point`, as a point is noted as 2 (or more) entries, leaving additional elements beyond longitude and latitude as intentionally undefined.

the *point* (pun intended) of my comment is to note that sometimes extending the value of a GeoJSON object is sometimes optimal, and am looking for a clean way of doing it, rather than being explicitly disallowed.

i think i already saw a couple of notes on the draft mailing list, but if you feel i should propose it, i am happy to.

adoyle commented:
I would cut it down to this:

--
# Extensibility

In GeoJSON, an "extension" is any property not defined in section 2.

## GeoJSON types are fixed

Extensions MUST NOT extend the fixed set ... and Geometry Collection.

## Semantics of GeoJSON properties are fixed

Extensions MUST NOT change the the semantics of GeoJSON properties.

## Processing (Non-normative)

Note that support for extensions can vary across implementations and no normative processing model for extensions is defined. Accordingly, implementations that rely too heavily on the use of extensions may experience reduced interoperability with other implementations.

Implementations that collect and aggregate GeoJSON objects from multiple sources may often need to deal with conflicting or incompatible extensions.

sgillies commented on considerations.mkd:
```diff
@@ -58,3 +58,38 @@ precision than necessary.
+contain extension properties SHOULD take sufficient care to ensure that
+extension properties used within the original document are preserved and
+serialized appropriately.
+
```
For a GeoJSON parser that extends a general purpose JSON parser, this is not a BFD. For a GeoJSON parser that is not a general purpose JSON parser (what were the developers thinking?), this might be challenging.

sgillies replied:
What about implementations that transform GeoJSON? https://github.com/geowurster/fio-buffer, for example.

I suggested on the list that applications which create new features can discard extensions freely. Those that modify features should take care to preserve. These guidelines, however, depend on a concept of feature identity that is not well defined for GeoJSON.

adoyle replied:
This is the slippery slope.

tschaub replied:
> For a GeoJSON parser that extends a general purpose JSON parser, this is not a BFD.

Can you add more detail here?  Re-serialization is not simply a matter of extending a JSON parser.  It means an implementation also needs to retain the original JSON-parsed object.

sgillies replied:
@tschaub what I mean is that if your GeoJSON parser extends, e.g. Python's `json.loads()`, you get the extension properties automatically and would have to go out of your way to remove them after parsing.

sgillies replied:
Reserialization is alarming folks and some of the implications brought up on the list by @martinthomson frighten me. I'm going to remove this.

I'm convinced that preserving extension properties is a helpful feature and that software that can pull it off will be appreciated, but not insistent on having the spec say anything about it.

sgillies commented on considerations.mkd:
```diff
@@ -58,3 +58,38 @@ precision than necessary.
+
+Implementations that collect and aggregate GeoJSON objects from multiple
+sources may often need to deal with conflicting or incompatible
+extensions.
```
This is just an acknowledgement of the problem. There's no one size fits all solution.

dret commented:
i like it as a starting point. should we create specific issues/PRs for removing some of the extensibility language that is part of other sections? to me it seems that having all of this in one place would be a an improvement.

mpdaly commented:
I would cut it down to this.
---

# Additional Members

The GeoJSON objects Feature and FeatureCollection, but not Point ... GeometryCollection, MAY include additional members for application- or domain-specific purposes.

Implementations that parse GeoJSON MUST NOT fail, or report errors, if such additional members are present.

sgillies commented:
@adoyle I eliminated the sections on re-serialization (unpopular and fraught with complications pointed out by @martinthomson) and aggregation (unnecessary). I am leaving the MUST NOT halt and catch fire bit in here. Before I will remove it, I think everyone needs to see an explanation of why an implementation needs to stop processing on encountering foreign properties.

@mpdaly why do you want a SHOULD NOT here?

> Implementations that parse GeoJSON SHOULD NOT fail, or report errors, if such additional members are present.

Why do you want the option to fail and report an error instead of just skipping over unrecognized foreign properties (members)? Why do you want other implementations to be able to fail instead of skipping?

mpdaly commented:
@sgillies My original wording was MUST NOT re unrecognised, but @adoyle has an aversion to such strict language except where absolutely necessary (as I understand it) and asked me to change it.

https://twitter.com/adoyle/status/669614093041577985

I'm guessing that he feels it unnecessary to actively prevent an implementation working however it sees fit (even if it is deemed unfriendly by the likes of us).

sgillies commented:
@jyutzler @adoyle a format specification certainly *can* put requirements on the behavior of processing applications. Atom does it: https://tools.ietf.org/html/rfc4287#section-6.

jyutzler commented:
Just because they did it doesn't mean they are right. It is pretty common for things to slip by without being noticed.

sgillies commented:
I've moved "Extending GeoJSON" up in the document and removed the MUST ignore sentence. I'm going to content myself with it already being covered in the I-JSON subsection of our "Interoperability considerations".

"Extending GeoJSON" is now entirely concerned with cautions and constraints for implementers and is more or less equivalent to @adoyle's suggestion at https://github.com/geojson/draft-geojson/pull/117#issuecomment-159702752.

adoyle commented:
+1 and thanks @sgillies
